keps/420-partial-admission/README.md

```diff
@@ -0,0 +1,173 @@
+
+### User Stories
+
+Kueue issue [420](https://github.com/kubernetes-sigs/kueue/issues/420) provides details on the initial feature details and its applicability for `batch/Job`.
+
+## Design Details
+
+### Workload API
+
+```go
+type PodSet struct {
+    // .......
+
+    // count is the number of pods for the spec.
+    // +kubebuilder:validation:Minimum=1
+    Count int32 `json:"count"`
+
+    // minimumCount is the minimum number of pods for the spec acceptable
+    // in case of partial admission.
+    //
+    // If not provided, partial admission for the current PodSet is not
+    // enabled.
+    // +optional
+    MinCount *int32 `json:"minCount,omitempty"`
+}
+
+```
+
+### Scheduler / Flavorassignment
+
+In case the workload proposed for the current scheduling cycle, does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find to find a lower counts combination that fits the available quota with or without borrowing.
+
+The search should be optimized (binary search) and preserve the proportion of pods lost across the variable count PodSets.
+
+The accepted number of pods in each PodSet are recorded in `workload.Status.Admission.PodSetAssignments[*].ResourceUsage.Count`
+
+In order to evaluate the potential success of the preemption, the preemption process should be split in:
+- Target selection, which should select the workloads within cohort that will need to be evicted, if the list is empty the assignment will be treated as `NoFit`. This step will take place during nomination.
+- Preemption issuing, will do the actual eviction of the targets previously selected.
+
+**Partial admission with multiple variable PodSets count**
+
+If multiple PodSets within a workload have variable counts the way the counts should be decrease is highly dependent on the job framework needs, currently the following is proposed:
+
+Starting with the PodSets:
+
+```go
+[]podSet { {Count: 1} , {Count: 4, MinCount: 2}, {Count: 20, MinCount: 10}}
+```
+
+And only being able to admit 19 pods, it should end up with
+
+
+```go
+[]podSet { {Count: 1} , {Count: 3}, {Count: 15}}
+```
+
+With both pods that accept partial admission losing 50% of their range.
+
+The presented solution is not taking into account which of the variable count PodSets is causing the "NoFit" status and can potentially decrease the count of PodSets that will otherwise fit. If this behaviour is not suitable for a specific job framework, the integration layer of the framework can take into account limiting the variable count PodSets to 1.
+
+### Jobframework
+
+```diff
+type GenericJob interface {
+    // ...
+
+	
+-    // RunWithNodeAffinity will inject the node affinity extracting from workload to job and unsuspend the job.
++    // RunWithPodSetsInfo will inject the node affinity and podsSet counts extracted from workload to the job and unsuspend the job.
+-    RunWithNodeAffinity(nodeSelectors []PodSetNodeSelector)
```

Comment:
Some thoughts: when designing the job framework, I was thinking we can define several interfaces for different Jobs, embedding all of them into the single GenericJob is hard to evolve especially we have more and more job types in the future if we have.